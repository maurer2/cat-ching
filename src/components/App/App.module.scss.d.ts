import globalClassNames from '../../style.d';
declare const classNames: typeof globalClassNames & {
  readonly container: 'container';
  readonly main: 'main';
};
export = classNames;
