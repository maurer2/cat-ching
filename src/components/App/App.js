import React, { useState } from 'react';
// import logo from '../../logo.svg';
import style from './App.module.css';

const getRandomAmount = () => {
  const number = Math.floor(Math.random() * 100) + 1;
  
  return number;
}

function App() {
  const [amount, setAmount] = useState(getRandomAmount());

  return (
    <div className={style.app}>
      <header className={style.header}>
        Header
      </header>
      <main className={style.main}>
        Main
      </main>
    </div>
  );
}

export default App;
