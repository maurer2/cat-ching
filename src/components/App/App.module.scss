.app {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
}

.header {
  padding: 1rem;
  text-align: center;
  background: black;
  color: white;
}

.main {
  display: flex;
  flex-grow: 1;
  background: white;
  color: black;
}

.form {
  display: contents;
}
