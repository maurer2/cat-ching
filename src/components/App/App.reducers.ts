import Money from '../../types/Money';
import { MoneyReducerState,
  MoneyReducerAction,
  moneyReducerActionTypes } from './App.types';

export const moneyReducer = (state: MoneyReducerState, { type, payload }: MoneyReducerAction) => {
  switch (type) {
    case moneyReducerActionTypes.ADD_TO_CURRENT_AMOUNT: {
      const currentAmount: Money = state.currentAmount.add(payload);
      return {
        ...state,
        currentAmount,
      };
    }
    case moneyReducerActionTypes.SUBTRACT_FROM_CURRENT_AMOUNT: {
      let currentAmount: Money = state.currentAmount.subtract(payload);

      if (currentAmount.isNegative()) {
        currentAmount = Money.fromNumber(0, 'GBP');
      }

      return {
        ...state,
        currentAmount,
      };
    }
    case moneyReducerActionTypes.RESET_STATE: {
      return {
        ...state,
        targetAmount: Money.fromRandom('GBP'),
        currentAmount: Money.fromNumber(0, 'GBP'),
      };
    }
    default: {
      return state;
    }
  }
};
