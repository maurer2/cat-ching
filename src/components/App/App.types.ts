import Money from '../../types/Money';
import Coin from '../../types/Coin';

export type AppProps = {
  coinList: ReadonlyArray<Coin>;
};

export type CoinsWithStableKeys = ReadonlyArray<{
  key: string;
  coin: Coin;
}>;

// taken and adapted for union types from
// https://www.newline.co/@bespoyasov/how-to-use-usereducer-with-typescript--3918a332
export type State = {
  targetAmount: Money;
  currentAmount: Money;
};

export const actionTypeValues = {
  ADD_TO_CURRENT_AMOUNT: 'ADD_TO_CURRENT_AMOUNT',
  SUBTRACT_FROM_CURRENT_AMOUNT: 'SUBTRACT_FROM_CURRENT_AMOUNT',
  RESET_STATE: 'RESET_STATE',
} as const;
export type ActionType = typeof actionTypeValues;
export type ActionTypeKeys = ActionType[keyof ActionType];

// lookup for payload types only
type ActionPayloads = {
  [actionTypeValues.ADD_TO_CURRENT_AMOUNT]: Money;
  [actionTypeValues.SUBTRACT_FROM_CURRENT_AMOUNT]: Money;
  [actionTypeValues.RESET_STATE]: never;
};

type ActionsWithoutPayload = Extract<ActionTypeKeys, 'RESET_STATE'>;
type ActionWithoutPayload = {
  type: ActionsWithoutPayload;
};

type ActionWithPayload<T extends ActionTypeKeys> = {
  // needs to also contain ActionsWithoutPayload
  type: T;
  payload: ActionPayloads[T];
};

export type Actions<T extends ActionTypeKeys> = T extends ActionsWithoutPayload
  ? ActionWithoutPayload
  : ActionWithPayload<T>;
