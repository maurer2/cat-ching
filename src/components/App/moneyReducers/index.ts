export { default } from './money.reducers';
