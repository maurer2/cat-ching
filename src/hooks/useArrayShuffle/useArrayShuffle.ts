import { useState, useEffect, useCallback } from 'react';

type ArrayWithRandomIndex<T> = {
  newIndex: number;
  data: T;
}[];

export default function useArrayShuffle<T>(
  sortedArray: ReadonlyArray<T>,
): readonly [ReadonlyArray<T>, () => void] {
  const [shuffledArray, setShuffledArray] = useState(sortedArray);

  const shuffle = useCallback(() => {
    // assign random number to each entry
    const arrayWithRandomNumbers: ArrayWithRandomIndex<T> = sortedArray.map((data) => ({
      newIndex: Math.random(),
      data,
    }));

    // sort by randomNumber
    arrayWithRandomNumbers.sort((entryA, entryZ) => (entryA.newIndex > entryZ.newIndex ? 1 : -1));

    // remove randomNumber again
    const newShuffledArray = arrayWithRandomNumbers.map(({ data }) => data);

    setShuffledArray(newShuffledArray);
  }, [sortedArray]);

  useEffect(() => {
    shuffle();
  }, [sortedArray, shuffle]);

  return [shuffledArray, shuffle] as const;
}
