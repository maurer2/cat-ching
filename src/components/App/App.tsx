import React, { useState, useReducer, ReducerWithoutAction, Reducer } from 'react';
import { shuffle } from 'lodash-es';

import Header from '../Header';
import Slider from '../Slider';
import Coin from '../Coin';
import Footer from '../Footer';
import Overlay from '../Overlay';

import Money from '../../types/Money';
import CoinType from '../../types/Coin';

import style from './App.module.scss';
import * as Types from './App.types';

function App({ coinList }: Types.AppProps): JSX.Element {
  const [coins, setCoins] = useReducer<ReducerWithoutAction<ReadonlyArray<CoinType>>>(
    (state) => shuffle(state),
    shuffle(coinList),
  );
  const [targetAmount, setTargetAmount] = useState<Money>(() => Money.fromRandom('GBP'));
  const [currentAmount, setCurrentAmount] = useReducer<Reducer<Money, Money>>((state, newState) => {
    const newCurrentAmount: Money = state.add(newState);

    if (newCurrentAmount.isNegative()) {
      return Money.fromNumber(0, 'GBP');
    }

    return newCurrentAmount;
  }, Money.fromNumber(0, 'GBP'));
  const overlayIsVisible: boolean = currentAmount.isEqualTo(targetAmount);

  function resetState(): void {
    setTargetAmount(Money.fromRandom('GBP'));
    setCurrentAmount(Money.fromNumber(0, 'GBP'));
    setCoins();
  }

  function handleSubmit(event): void {
    event.preventDefault();
  }

  return (
    <div className={style.container}>
      <Header
        targetAmount={targetAmount}
        handleReset={resetState}
      />
      <main className={style.main}>
        <form
          className={style.form}
          onSubmit={handleSubmit}
          action=""
          method=""
        >
          <Slider key={targetAmount.valueAsFormattedString}>
            {coins.map((coin) => (
              <Coin
                key={coin.name}
                name={coin.name}
                image={coin.image}
                amount={coin.amount}
                handleAmountChange={setCurrentAmount}
                size={coin.size}
              />
            ))}
          </Slider>
        </form>
      </main>
      <Footer currentAmount={currentAmount} />
      {overlayIsVisible && <Overlay />}
    </div>
  );
}

export default App;
