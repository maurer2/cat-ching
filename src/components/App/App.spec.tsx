import React from 'react';
import { render } from '@testing-library/react';

import Component from './App';
import * as Types from './App.types';
import Money from '../../types/Money';
import useArrayShuffle from '../../hooks/useArrayShuffle';

const mockedCoinList = [
  {
    name: '1 Pound',
    image: '1l.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(100, 'GBP'),
  },
  {
    name: '2 Pounds',
    image: '2l.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(200, 'GBP'),
  },
  {
    name: '1 Penny',
    image: '1p.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(1, 'GBP'),
  },
  {
    name: '2 Pence',
    image: '2p.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(1, 'GBP'),
  },
  {
    name: '5 Pence',
    image: '5p.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(1, 'GBP'),
  },
  {
    name: '10 Pence',
    image: '10p.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(1, 'GBP'),
  },
  {
    name: '20 Pence',
    image: '20p.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(1, 'GBP'),
  },
  {
    name: '50 Pence',
    image: '50p.png',
    size: {
      height: 1,
      unit: 'mm',
      width: 1,
    },
    amount: Money.fromNumber(1, 'GBP'),
  },
] as const;
vi.mock('../../hooks/useArrayShuffle', () => ({
  default: vi.fn(),
}));

describe('AppContainer', () => {
  beforeEach(() => {
    // set fixed random target value
    vi.spyOn(global.Math, 'random').mockReturnValueOnce(0.5);
    // https://github.com/vitest-dev/vitest/issues/745
    vi.mocked(useArrayShuffle).mockReturnValue([mockedCoinList, vi.fn()]);
  });

  afterEach(() => {
    vi.spyOn(global.Math, 'random').mockRestore();
  });

  const propsDefault: Types.AppProps = {
    coinList: mockedCoinList,
  };

  const setup = (props = {}) => render(<Component {...propsDefault} {...props} />);

  it('renders', () => {
    const { getByTestId } = setup();

    expect(getByTestId('app')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });
});
