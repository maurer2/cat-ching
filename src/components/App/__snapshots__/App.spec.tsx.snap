// Vitest Snapshot v1

exports[`AppContainer > matches snapshot 1`] = `
<div>
  <div
    class="_container_fdx0g_1"
    data-testid="app"
  >
    <header
      class="_header_q2xj4_1"
      data-testid="header"
    >
      <h1
        class="_title_q2xj4_11"
      >
        Cat-Ching
      </h1>
      <dl
        class="_detail_q2xj4_18"
      >
        <dt
          class="_detailKey_q2xj4_28"
        >
          Target amount:
        </dt>
        <dd
          class="_detailValue_q2xj4_32"
        >
          Â£5.01
        </dd>
      </dl>
      <button
        class="_resetButton_q2xj4_37"
        type="button"
      >
        Reset
      </button>
    </header>
    <main
      class="_main_fdx0g_8"
    >
      <ul
        class="_slider_xf5j5_1"
        data-testid="slider"
      >
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":r8:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                1 Pound
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/1l.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":r9:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                2 Pounds
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/2l.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":ra:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                1 Penny
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/1p.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":rb:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                2 Pence
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/2p.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":rc:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                5 Pence
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/5p.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":rd:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                10 Pence
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/10p.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":re:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                20 Pence
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/20p.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
        <li
          class="_slide_xf5j5_1"
        >
          <div
            class="_container_r1k0e_1"
            data-testid="coin"
          >
            <label
              class="_header_r1k0e_5"
              for=":rf:"
            >
              <span
                class="_title--hidden_r1k0e_16"
              >
                50 Pence
              </span>
              <img
                alt="Coin"
                class="_image_r1k0e_22"
                src="/images/50p.png"
                style="width: 3.52113%;"
              />
            </label>
            <div
              class="_buttonGroup_r1k0e_28"
            >
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Add amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Subtract amount
              </button>
              <button
                class="_button_r1k0e_28"
                type="button"
              >
                Show hint
              </button>
            </div>
          </div>
        </li>
      </ul>
    </main>
    <footer
      class="_footer_1j6yp_1"
      data-testid="footer"
    >
      <dl
        class="_detail_1j6yp_9"
      >
        <dt
          class="_detailKey_1j6yp_19"
        >
          Current amount:
        </dt>
        <dd
          class="_detailValue_1j6yp_24"
        >
          Â£0.00
        </dd>
      </dl>
      <button
        class="_toggleButton_1j6yp_29"
        type="button"
      >
        Hide current amount
      </button>
    </footer>
  </div>
</div>
`;
